src/key-value-engine/B-Tree_vs_The_Log-Structured_Merge-Tree.md

```diff
@@ -0,0 +1,101 @@
+
+### Write Amplification
+
+`Write amplification` is the ratio of the amount of data written to the storage medium versus the amount of data written to the database. 
+
+For example, if you are writing 10 MB to the database and you observe 30 MB disk write rate, your write amplification is 3.
+
+Flash memory and solid-state drives (SSDs) can be written to only a finite number of times, so that write amplification will decrease the flash lifetime.
+
+There is another write amplification phenomenon associated with the flash memory and SSDs because flash memory must be erased before it can be rewritten.
+
+### Read Amplification
+
+`Read amplification` is the number of disk reads per query. 
+
+For example, if you need to read 5 pages to answer a query, read amplification is 5.
+
+### Space Amplification
+
+`Space amplification` is the ratio of the amount of data on the storage medium versus the amount of data on the database. 
+
+For example, if you put 10MB in the database and this database uses 100MB on the disk, then the space amplification is 10.
+
+## Analysis
+
+The term B-tree may refer to a specific design or a general class of designs. In the narrow sense, a B-tree stores keys in its internal nodes but need not store those keys in the records at the leaves. The [B+ tree](https://en.wikipedia.org/wiki/B%2B_tree#Insertion) is one of the most famous variations of B-tree, behind which the idea is that internal nodes only contain keys, and to which an additional level which contains values is added at the bottom with linked leaves.
+
+LSM-tree performs `compaction` to merge several `sstable`s into one new `sstable` which contains only the live data from the input `sstable`s. Compaction helps LSM-tree to recycle space and reduce read amplification. There are two kinds of `compaction strategy` which is `Size-Tiered compaction strategy` (STCS) and `Level-Based compaction strategy` (LBCS). The idea behind STCS is compacting small `sstable`s into medium `sstable`s when LSM-tree has enough small `sstable`s and compacting medium `sstable`s into large `sstable`s when LSM-tree has enough medium `sstable`s. The idea of LBCS is to organize data into levels which contains one sorted run..
+
+This section discusses about the write amplification and read amplification of B+tree and Level-Based LSM-tree. 
+
+### B+ Tree
+
+In the B+ tree, copies of the keys are stored in the internal nodes; the keys and records are stored in leaves; in addition, a leaf node may include a pointer to the next leaf node to increase sequential access performance.
+
+To simplify the analysis, assume that the block size of the tree is `B`, and keys, pointers, and records are constant size, so that each internal node contains `O(B)` children and each leaf contains `O(B)` data records. (The root node is a special case, and can be nearly empty in some situations.) Under all these assumptions, the depth of a B tree is 
+
+<center><code>O(log<sub>B</sub>N/B)</code></center>
```

Comment:
don't use HTML, use Fenced code blocks, https://help.github.com/articles/creating-and-highlighting-code-blocks/

Reply:
I need to use HTML <sub> to support subscripting formula, but seems Fenced code blocks doesn't work well with it. So it's a work around.

Reply:
If we are going to read this with gitbook, I suggest we can just rely on gitbook to render the equations.
https://gitbookio.gitbooks.io/documentation/content/format/math.html

Reply:
How about mdbook @Hoverbear

Reply:
another simple way may use http://www.codecogs.com to generate an url link for your formula

Reply:
We can use KaTeX by overriding the header in the theme. After that PR is merged you can just use LaTeX as detailed in: https://github.com/rust-lang-nursery/mdBook/issues/49#issuecomment-167002731

I'll make a PR.

Reply:
See https://github.com/tikv/deep-dive-tikv/pull/9